sys/linux/rdma.txt

```diff
@@ -0,0 +1,1464 @@
+	mask		ib_flow_esp_filter
+}
+
+ib_flow_tunnel_filter {
+	tunnel_id	int32
+}
+
+ib_flow_spec_tunnel {
+	type		const[IB_FLOW_SPEC_VXLAN_TUNNEL, int32]
+	size		const[0x10, int16]
+	reserved	const[0x0, int16]
+	val		ib_flow_tunnel_filter
+	mask		ib_flow_tunnel_filter
+}
+
+ib_flow_gre_filter {
+	c_ks_res0_ver	int16
+	protocol	int16
+	key		int32
+}
+
+ib_flow_spec_gre {
+	type		const[IB_FLOW_SPEC_GRE, int32]
+	size		const[0x18, int16]
+	reserved	const[0x0, int16]
+	val		ib_flow_gre_filter
+	mask		ib_flow_gre_filter
+}
+
+ib_flow_mpls_filter {
+	tag	int32
+}
+
+ib_flow_spec_mpls {
+	type		const[IB_FLOW_SPEC_MPLS, int32]
+	size		const[0x10, int16]
+	reserved	const[0x0, int16]
+	val		ib_flow_mpls_filter
+	mask		ib_flow_mpls_filter
+}
+
+ib_flow_spec_action_tag {
+	type		const[IB_FLOW_SPEC_ACTION_TAG, int32]
+	size		const[0x10, int16]
+	reserved	const[0x0, int16]
+	tag_id		int32
+	reserved2	const[0x0, int32]
+}
+
+ib_flow_spec_action_drop {
+	type		const[IB_FLOW_SPEC_ACTION_DROP, int32]
+	size		const[0x8, int16]
+	reserved	const[0x0, int16]
+}
+
+ib_uverbs_flow_spec_action_count {
+	type		const[IB_FLOW_SPEC_ACTION_COUNT, int32]
+	size		const[0x10, int16]
+	reserved0	const[0x0, int16]
+	handle		vcontext_handle
+	reserved1	const[0x0, int32]
+}
+
+union_ib_flow_spec [
+	eth	ib_flow_spec_eth
+	ipv6	ib_flow_spec_ipv6
+	ipv4	ib_flow_spec_ipv4
+	tcp_udp	ib_flow_spec_tcp_udp
+	tunnel	ib_flow_spec_tunnel
+	gre	ib_flow_spec_gre
+	esp	ib_flow_spec_esp
+	mpls	ib_flow_spec_mpls
+	tag	ib_flow_spec_action_tag
+	drop	ib_flow_spec_action_drop
+	count	ib_uverbs_flow_spec_action_count
+]
+
+ib_uverbs_flow_attr {
+	type		int32[0:3]
+	size		bytesize[flow_specs, int16]
+	priority	int16[0:7]
+	num_of_specs	len[flow_specs, int8]
+	reserved1	const[0x0, int8]
+	reserved2	const[0x0, int8]
+	port		int8
+	flags		flags[ib_flow_flags, int32]
+	flow_specs	array[union_ib_flow_spec]
+}
+
+create_flow_cmd_ex {
+	comp_mask	const[0x0, int32]
+	qp_handle	qp_handle
+	flow_attr	ib_uverbs_flow_attr
+}
+
+create_flow_resp {
+	comp_mask	int32
+	flow_handle	flow_handle
+}
+
+create_flow_cmd {
+	command			const[IB_USER_VERBS_EX_CMD_CREATE_FLOW, int32]
+	in_words		bytesize8[flow_ex, int16]
+	out_words		const[0x1, int16]
+	response		ptr[out, create_flow_resp]
+	provider_in_words	const[0x0, int16]
+	provider_out_words	const[0x0, int16]
+	reserved		const[0x0, int32]
+	flow_ex			create_flow_cmd_ex
+}
+
+destroy_flow_cmd {
+	command			const[IB_USER_VERBS_EX_CMD_DESTROY_FLOW, int32]
+	in_words		const[0x6, int16]
+	out_words		const[0x0, int16]
+	response		const[0x0, int64]
+	provider_in_words	const[0x0, int16]
+	provider_out_words	const[0x0, int16]
+	reserved		const[0x0, int32]
+	comp_mask		const[0x0, int32]
+	flow_handle		flow_handle
+}
+
+# ioctl interface
+
+ib_uverbs_attr_flags = UVERBS_ATTR_F_MANDATORY
+ib_read_counters_flags = IB_UVERBS_READ_COUNTERS_PREFER_CACHED
+resource vcontext_handle[int32]
+
+ib_uverbs_create_counters_cmd {
+	length		bytesize8[ib_uverbs_create_counters_cmd, int16]
```

Comment:
It seems all common fields for all ioctl's correspond to `ib_uverbs_ioctl_hdr` struct.
Please define it explicitly using templates similarly to the commands for write.